[1.18.x] Backport grindstone event from 1.19

KnightMiner opened the pull request:
Backport of #8934
Contains test mod from #8578
Includes fix for event from #9089

ferriarnus committed:
Backport grindstone event from 1.19

KnightMiner committed:
Fix several bugs in the original PR related to canceling the event

Fix canceling the event not clearing the result item, causing it to dupe whatever the result with the previous item was
Fix canceling the event running vanilla behavior if vanilla behavior ran at all since the menu was opened and no other listener changed the XP

KnightMiner committed:
Add comments to the test mod explaining the expected behavior

KnightMiner committed:
Clarify docs on behavior of the XP

Makes it clear XP is ignored in on place if the output is empty or the event is canceled

KnightMiner committed:
Fix broadcast changes not being called if the event overrides grindstone behavior

Not sure its strictly needed, but every vanilla branch in the function calls it and its a trivial change.

KnightMiner committed:
Add enabled flag to the test mod

KnightMiner committed:
Improve javadocs

(cherry picked from commit e12c849cc0f4e1a031c06be9cb6489d21e4b1685)

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,175 @@
+     * </ul>
+     */
+    @Cancelable
+    public static class OnplaceItem extends GrindstoneEvent
```
Please fix the casing of the class name

KnightMiner replied:
This is the casing used in 1.19. I know its erronious, but I think consistency is better than forcing people updating from 1.18 to 1.19 to switch to a worse name.

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,175 @@
+
+    /**
+     * This event is {@link Cancelable} <br>
+     * {@link OnTakeItem} is fired when the output in a grindstone are is taken. <br>
```
```suggestion
     * {@link OnTakeItem} is fired when the output of a grindstone is taken. <br>
```

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    /**
+     * This event is {@link Cancelable} <br>
```
```suggestion
     * This event is {@link Cancelable}.<br>
```

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,175 @@
+     * {@link OnTakeItem} is fired when the output in a grindstone are is taken. <br>
+     * It is called from {@link GrindstoneMenu#GrindstoneMenu(int, Inventory)}. <br>
+     * If the event is canceled, vanilla behavior will not run, and no inputs will be consumed. <br>
+     * if the amount of experience is larger than or equal 0, the vanilla behavior for calculating experience will not run. <br>
```
```suggestion
     * if the amount of experience is larger than or equal to 0, the vanilla behavior for calculating experience will not be used. <br>
```

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,175 @@
+    /**
+     * This event is {@link Cancelable} <br>
+     * {@link OnTakeItem} is fired when the output in a grindstone are is taken. <br>
+     * It is called from {@link GrindstoneMenu#GrindstoneMenu(int, Inventory)}. <br>
```
```suggestion
     * The event is called from {@link GrindstoneMenu#GrindstoneMenu(int, Inventory)}. <br>
```

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,175 @@
+import net.minecraftforge.eventbus.api.Cancelable;
+import net.minecraftforge.eventbus.api.Event;
+
+public abstract class GrindstoneEvent extends Event
```
The getters in this class (and children) should use inline `{@return}`s. Additionally, the `GrindstoneEvent` should at least mention the place and take events. (using `@see`)

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/GrindstoneEventTest.java:
```diff
@@ -0,0 +1,86 @@
+import net.minecraftforge.fml.common.Mod;
+
+@Mod("grindstone_event_test")
+public class GrindstoneEventTest {
```
Brace on new line.

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/GrindstoneEventTest.java:
```diff
@@ -0,0 +1,86 @@
+@Mod("grindstone_event_test")
+public class GrindstoneEventTest {
+    private static final boolean ENABLED = false;
+    public GrindstoneEventTest() {
```
Brace on new line.

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/GrindstoneEventTest.java:
```diff
@@ -0,0 +1,86 @@
+public class GrindstoneEventTest {
+    private static final boolean ENABLED = false;
+    public GrindstoneEventTest() {
+        if (ENABLED) {
```
Brace on new line.

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/GrindstoneEventTest.java:
```diff
@@ -0,0 +1,86 @@
+            event.setNewTopItem(top);
+        }
+    }
+}
```
LF at EOF.

ChampionAsh5357 reviewed (changes requested):
Some formatting fixes missed in the 1.19 version ~~looks at maty~~.

Matyrobbrt reviewed (changes requested):
Tested the test mod, and it works, I've got just a few nitpicks about javadocs and grammar.

KnightMiner committed:
Formatting fixes

KnightMiner committed:
Apply javadoc suggestions from review

Co-authored-by: matyrobbrt <65940752+Matyrobbrt@users.noreply.github.com>

KnightMiner committed:
Correct capitalization in event name

KnightMiner committed:
Javadoc improvements

Remove some redundancy in certain javadoc comments
Add @see to the header
Inline single line javadocs

KnightMiner committed:
Add some funny curly brackets next to returns

Seems the kids these days have invented a new way to write return statements in javadocs

KnightMiner committed:
Remove extra dots and capital letters

The return tag apparently puts more dots at the end, and having two dots on each sentence is not in the budget. Have also reached the capital letter quota

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,160 @@
+        return bottom;
+    }
+
+    /** {@return the experience amount given to the player. It will be {@code -1} unless {@link #setXp(int)} is called} */
```
```suggestion
    /** {@return the experience amount given to the player} It will be {@code -1} unless {@link #setXp(int)} is called. */
```

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -0,0 +1,160 @@
+     * {@link OnTakeItem} is fired when the output of a grindstone is taken. <br>
+     * The event is called from {@link GrindstoneMenu#GrindstoneMenu(int, Inventory)}. <br>
+     * If the event is canceled, vanilla behavior will not run, and no inputs will be consumed. <br>
+     * if the amount of experience is larger than or equal to 0, the vanilla behavior for calculating experience will not be used. <br>
```
```suggestion
     * If the amount of experience is larger than or equal to 0, the vanilla behavior for calculating experience will not be used. <br>
```

KnightMiner committed:
Apply suggestions from code review

Co-authored-by: matyrobbrt <65940752+Matyrobbrt@users.noreply.github.com>
